README.md
<p align="center">
    <img width="400" height="125" src="https://raw.githubusercontent.com/elgentos/masquerade/master/art/logo.png" alt="Masquerade logo" />
</p>

# Masquerade

## Faker-driven, platform-agnostic, locale-compatible data faker tool

Point Masquerade to a database, give it a rule-set defined in YAML and Masquerade will anonymize the data for you
 automatically!

<img src="https://user-images.githubusercontent.com/431360/42574650-30e8d186-851f-11e8-9693-c23b426c43f2.png" width="600" />

### Out-of-the-box supported frameworks

- Magento 2
- Shopware 6

### Customization

You can add your own configuration files in a directory named `config` in the same directory as where you run masquerade. The configuration files will be merged with any already present configuration files for that platform, overriding any out-of-the-box values.

See the [Magento 2 YAML files](https://github.com/elgentos/masquerade/tree/master/src/config/magento2) as examples for notation.

For example, to override the `admin.yaml` for Magento 2, you place a file in `config/magento2/admin.yaml`. For example, if you want to completely disable/skip a group, just add this content;

```
admin:
```

You can add your own config files for custom tables or tables from 3rd party vendors. Here are a few examples:
- [Magento 2 Aheadworks YAML file](https://github.com/elgentos/masquerade/wiki/%5BMagento-2%5D-Aheadworks-YAML-file)
- [Magento 2 Amasty YAML file](https://github.com/elgentos/masquerade/wiki/%5BMagento-2%5D-Amasty-YAML-file)
- [Shopware 6 Frosh YAML file](https://github.com/elgentos/masquerade/wiki/%5BShopware-6%5D-Frosh-YAML-file)

To generate such files, you can run the `masquerade identify` command. This will look for columns that show a hint of personal identifiable data in the name, such as `name` or `address`. It will interactively ask you to add it to a config file for the chosen platform.

### Partial anonymization

You can affect only certain records by including a 'where' clause - for example to avoid anonymising certain admin accounts, or to preserve data used in unit tests, like this:

```yaml
customers:
  customer_entity:
    provider: # this sets options specific to the type of table
      where: "`email` not like '%@mycompany.com'" # leave mycompany.com emails alone
```

### Delete Data

You might want to fully or partially delete data - eg. if your developers don't need sales orders, or you want to keep the database size a lot smaller than the production database.  Specify the 'delete' option.

When deleting some Magento data, eg. sales orders, add the command line option `--with-integrity` which enforces foreign key checks, so for example sales\_invoice records will be deleted automatically if their parent sales\_order is deleted:

```yaml
orders:
  sales_order:
    provider:
      delete: true
      where: "customer_id != 3" # delete all except customer 3's orders because we use that for testing
    # no need to specify columns if you're using 'delete'      
```

If you use 'delete' without a 'where', and without '--with-integrity', it will use 'truncate' to delete the entire table.  It will not use truncate if --with-integrity is specified since that bypasses key checks.

### Magento EAV Attributes

You can use the Magento2Eav table type to treat EAV attributes just like normal columns, eg.

```yaml
products:
  catalog_product_entity: # specify the base table of the entity
    eav: true
    provider:
      where: "sku != 'TESTPRODUCT'" # you can still use 'where' and 'delete'
    columns:
      my_custom_attribute:
        formatter: sentence
      my_other_attribute:
        formatter: email

  catalog_category_entity:
    eav: true
    columns:
      description: # refer to EAV attributes like normal columns
        formatter: paragraph

```

### Formatter Options

For formatters, you can use all default [Faker formatters](https://github.com/fzaninotto/Faker#formatters).

#### Custom Data Providers / Formatters

You can also create your own custom providers with formatters. They need to extend `Faker\Provider\Base` and they need to live in either `~/.masquerade` or `.masquerade` relative from where you run masquerade.

An example file `.masquerade/Custom/WoopFormatter.php`;

```php
<?php

namespace Custom;

use Faker\Provider\Base;

class WoopFormatter extends Base {

    public function woopwoop() {
        $woops = ['woop', 'wop', 'wopwop', 'woopwoop'];
        return $woops[array_rand($woops)];
    }
}
```

And then use it in your YAML file. A provider needs to be set on the column name level, not on the formatter level.

```
customer:
  customer_entity:
    columns:
      firstname:
        provider: \Custom\WoopFormatter
        formatter:
          name: woopwoop
```

### Custom Table Type Providers

Some systems have linked tables containing related data - eg. Magento's EAV system, Drupal's entity fields and Wordpress's post metadata tables.  You can provide custom table types. 
In order to do it you need to implement 2 interfaces:
 - `Elgentos\Masquerade\DataProcessorFactory` is to instantiate your custom processor. It receives table service factory, output object and whole array of yaml configuration specified for your table.
 - `Elgentos\Masquerade\DataProcessor` is to process various operations required by run command like:

   - `truncate` should truncate table in provided table via configuration
   - `delete` should delete table in provided table via configuration
   - `updateTable` should update table with values provided by generator based on columns definitions in the configuration. 
     See `Elgentos\Masquerade\DataProcessor\RegularTableProcessor::updateTable` for a reference.


First you need to start with a factory that will instantiate an actual processor

An example file `.masquerade/Custom/WoopTableFactory.php`;
```php
<?php

namespace Custom;

use Elgentos\Masquerade\DataProcessor;
use Elgentos\Masquerade\DataProcessor\TableServiceFactory;
use Elgentos\Masquerade\DataProcessorFactory;
use Elgentos\Masquerade\Output;
 
class WoopTableFactory implements DataProcessorFactory 
{

    public function create(
        Output $output, 
        TableServiceFactory $tableServiceFactory,
        array $tableConfiguration
    ): DataProcessor {
        $tableService = $tableServiceFactory->create($tableConfiguration['name']);

        return new WoopTable($output, $tableService, $tableConfiguration);
    }
}
```

An example file `.masquerade/Custom/WoopTable.php`;

```php
<?php

namespace Custom;

use Elgentos\Masquerade\DataProcessor;
use Elgentos\Masquerade\DataProcessor\TableService;
use Elgentos\Masquerade\Output;

class WoopTable implements DataProcessor
{
    /** @var Output */
    private $output;

    /** @var array */
    private $configuration;

    /** @var TableService */
    private $tableService;

    public function __construct(Output $output, TableService $tableService, array $configuration)
    {
        $this->output = $output;
        $this->tableService = $tableService;
        $this->configuration = $configuration;
    }

    public function truncate(): void
    {
        $this->tableService->truncate();
    }
    
    public function delete(): void
    {
        $this->tableService->delete($this->configuration['provider']['where'] ?? '');
    }
    
    public function updateTable(int $batchSize, callable $generator): void
    {
        $columns = $this->tableService->filterColumns($this->configuration['columns'] ?? []);
        $primaryKey = $this->configuration['pk'] ?? $this->tableService->getPrimaryKey();
        
        $this->tableService->updateTable(
            $columns, 
            $this->configuration['provider']['where'] ?? '', 
            $primaryKey,
            $this->output,
            $generator,
            $batchSize
        );
    }
}
```

And then use it in your YAML file. A processor factory needs to be set on the table level, and can be a simple class name, or a set of options which are available to your class.

```yaml
customer:
  customer_entity:
    processor_factory: \Custom\WoopTableFactory
    some_custom_config:
      option1: "test"
      option2: false
    columns:
      firstname:
        formatter:
          name: firstName
```

### Installation

Download the phar file:

```
curl -L -o masquerade.phar https://github.com/elgentos/masquerade/releases/latest/download/masquerade.phar
```

### Usage

```
$ php masquerade.phar run --help

Description:
  List of tables (and columns) to be faked

Usage:
  run [options]

Options:
      --platform[=PLATFORM]
      --driver[=DRIVER]      Database driver [mysql]
      --database[=DATABASE]
      --username[=USERNAME]
      --password[=PASSWORD]
      --host[=HOST]          Database host [localhost]
      --port[=PORT]          Database port [3306]
      --prefix[=PREFIX]      Database prefix [empty]
      --locale[=LOCALE]      Locale for Faker data [en_US]
      --group[=GROUP]        Comma-separated groups to run masquerade on [all]
      --with-integrity       Run with foreign key checks enabled
      --batch-size=BATCH-SIZE  Batch size to use for anonymization [default: 500]
```

You can also set these variables in a `config.yaml` file in the same location as where you run masquerade from, for example:

```yaml
platform: magento2
database: dbnamehere
username: userhere
password: passhere
host: localhost
port: porthere
```

### Running it nightly

Check out the wiki on how to run Masquerade nightly in CI/CD;
- [Usage with Gitlab CI/CD](https://github.com/elgentos/masquerade/wiki/Usage-with-Gitlab-CI-CD)
- [Usage with Github Actions](https://github.com/elgentos/masquerade/wiki/Usage-with-Github-Actions)
- [Usage with Bitbucket Pipelines](https://github.com/elgentos/masquerade/wiki/Usage-with-Bitbucket-Pipelines)

### Building from source

To build the phar from source you can use the `build.sh` script. Note that it depends on [Box](https://github.com/box-project/box) which is included in this repository.

```
# git clone https://github.com/elgentos/masquerade
# cd masquerade
# composer install
# chmod +x build.sh
# ./build.sh
# bin/masquerade
```

### Debian Packaging

To build a deb for this project run:

```
# apt-get install debhelper cowbuilder git-buildpackage
# export ARCH=amd64
# export DIST=buster
# cowbuilder --create --distribution buster --architecture amd64 --basepath /var/cache/pbuilder/base-$DIST-amd64.cow --mirror http://ftp.debian.org/debian/ --components=main
# echo "USENETWORK=yes" > ~/.pbuilderrc
# git clone https://github.com/elgentos/masquerade
# cd masquerade
# gbp buildpackage --git-pbuilder --git-dist=$DIST --git-arch=$ARCH --git-ignore-branch -us -uc -sa --git-ignore-new
```

To generate a new `debian/changelog` for a new release:
```
export BRANCH=master
export VERSION=$(date "+%Y%m%d.%H%M%S")
gbp dch --debian-tag="%(version)s" --new-version=$VERSION --debian-branch $BRANCH --release --commit
```

#### Credits

- Built by [elgentos](https://github.com/elgentos)
- Logo by [Caneco](https://twitter.com/caneco)


debian/changelog
masquerade (20221031.111417) UNRELEASED; urgency=medium

  [ peterjaap ]
  * Added yaml files for Magento1 - not correct because it has no EAV support yet!

  [ Rick van de Loo ]
  * add debian packaging (#7)

  [ peterjaap ]
  * Cool Carbon screenshot which is all the hype now - closes #14

  [ Jan Knipper ]
  * Added charset to command line options (#15)

  [ peterjaap ]
  * If formatter is unknown, use null for the value in order to remove personal data even though we don't have a perfect substitute - closes #2
  * Added extra level in yaml for formatter, to add options
  * Moved optional, unique and nullColumnBeforeRun one level deeper since it belongs to the column, not the formatter
  * Set randomNumber to 8 digits and updated RunCommand to expand options into method parameters
  * Allow for simple and extended syntax for formatter
  * Implementing chunking of database retrieval
  * New version
  * Added fixed 'formatter' - closes #9
  * Updated dist file
  * Built in checks to see if table and column exist
  * Built in checks to see if table and column exist, closes #11
  * Added masquerade dist
  * First code to include custom Faker providers
  * Update dist
  * Removed hardcoded files with a scandir
  * Added test to see if loading actually works - now on to an autoloader
  * Fixed config file loading for phar and non-phar
  * Added option to supply Faker providers with formatters
  * Updated README and dist
  * Moved custom autoloader to custom Application class, changed order of path prefixes to look in and added break in foreach loop to prevent class already defined errors
  * Removed duplicate declaration of isPhar()
  * UPdated dist
  * Fixed list command to also display formatter parameters
  * Changed bulid process to Box, added .php71 stopfile, updated smaller dist (11M > 3M), closes #12
  * Updated build script and README
  * Added illuminate/support, removed config file from dist
  * Refactored ListCommand int GroupsCommand to allow listing when no arguments are given (closes #26) and placed logo and version at top of help commands too
  * Upped version to 0.1.1
  * Replaced hassankhan/config with elgentos/parser (closes 22) Created configHelper to help with elgentos/parser Updated version to 0.1.2
  * Added LICENSE & updated dist
  * Removed dist file from repo, we'll place it under releases from now on
  * Updated download URL
  * Removed chmod command, Box does that for us
  * Place nullifying outside chunking
  * Ran phpstan level 7 and fixed errors
  * Added Identify Command
  * Raised version to 0.1.3
  * Removed test config.yaml and added it to gitignore
  * Made config.yaml also available in all config dirs
  * Cut off example values that are longer than 30 chars
  * Upped version to 0.1.4
  * Added rating yaml to magento2 source files
  * Fixed uniqueness of fields - closing #25
  * Upped to version 0.1.5

  [ Jeroen Boersma ]
  * Updated to latest elgentos/parser:2.4.0

  [ Michiel Gerritsen ]
  * Show a warning when the correct PHP version is not available

  [ peterjaap ]
  * Raised version to 0.1.7 to create new release with phar file
  * Add .idea to .gitignore
  * Changed precedence order for configuration settings; inline settings now take precedence over config file, as it should
  * Fix bug where Elgentos/Parser threw warning when parsing config YAML dir
  * Removed firstname from customer_grid_flat table configuration since it does not exist
  * Fixed precedence for prefix, locale and platformName, raised version to 0.1.9

  [ Timon de Groot ]
  * Fix undefined index: prefix
  * Update elgentos/parser (2.4.0 => 2.6.1)
  * Fix remaining undefined index notices
  * Add xdg config path to possible configDirs
  * Add Magento Enterprise tables
  * Add newlines

  [ Erik Hansen ]
  * Update readme with link that will always point to latest phar

  [ Peter Jaap Blaakmeer ]
  * Update README.md
  * Update README.md
  * Update README.md

  [ Caneco ]
  * improvement: add brand new logo to the project

  [ Peter Jaap Blaakmeer ]
  * Apply fixes from StyleCI

  [ Erik Hansen ]
  * Update installation command to be more broadly supported
  * Add readme making it obvious that M1 is not supported

  [ Peter Jaap Blaakmeer ]
  * Updated IP formatters for rating table
  * Raised version to 0.1.12

  [ Tjitse ]
  * Use random number in remote_ip_long
  * Fix remote_ip_long formatter

  [ John ORourke ]
  * updated README to clarify --group syntax
  * WIP on allowing different types of linked/EAV tables

  [ John ]
  * WIP - Simple table type now works

  [ John ORourke ]
  * styleci fixes
  * updated README and added config samples for delete/where options and custom table types
  * fixed build.sh to work with the default box installer process

  [ John ]
  * added CLI option to specify a custom config folder
  * allow table config "provider" option to be an option array with optional "class"
  * Tested and working with Magento 2.3.4, Amasty Order Attributes, and regular tables
  * non-functional changes: style fixes and optimisation
  * update to allow multiple --config options eg. --config=./global --config=./local
  * Added sensible defaults for primary key guessing; Simple table type now allows subclass to override orderBy; support multiple EAV attributes of the same type; warn if using WHERE on a nulled column
  * Get accurate count for the progress bar even when using groupBy due to per-store values

  [ Tjitse Efdé ]
  * Add option to specify database port
  * Add port option to readme, set default port to 3306

  [ Peter Jaap Blaakmeer ]
  * Updated illuminate/database updated from v5.8.35 to v6.20.13 and accompanying packages
  * Raised version to 0.1.13

  [ John ]
  * separate delete/truncate options to allow foreign key constraints to be met
  * added example magento2 configs with sales orders deleted
  * illuminate 6.20 removed array_get/array_first helpers - replace with Arr::get Arr::first

src/Elgentos/Masquerade/Console/ExportCommand.php
<?php

namespace Elgentos\Masquerade\Console;

use Elgentos\Masquerade\DataProcessor\DefaultDataProcessorFactory;
use Elgentos\Masquerade\DataProcessor\TableDoesNotExistsException;
use Elgentos\Masquerade\DataProcessor\TableServiceFactory;
use Elgentos\Masquerade\DataProcessorFactory;
use Elgentos\Masquerade\Helper\Config;
use Elgentos\Masquerade\Output;
use Faker\Generator;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;
use Faker\Factory as FakerFactory;
use Symfony\Component\Console\Helper\ProgressBar;
use Illuminate\Database\Capsule\Manager as Capsule;
use Illuminate\Support\Arr;

class ExportCommand extends Command
{
    const LOGO = '
._ _  _. _ _.    _ .__. _| _
| | |(_|_>(_||_|(/_|(_|(_|(/_
            |
                   by elgentos';

    const VERSION = '1.0.0';

    const DEFAULT_DATA_PROCESSOR_FACTORY = DefaultDataProcessorFactory::class;

    protected $config;

    /**
     * @var InputInterface
     */
    protected $input;

    /**
     * @var Output
     */
    protected $output;

    protected $platformName;
    protected $locale;

    /**
     * @var \Illuminate\Database\Connection
     */
    protected $db;
    protected $group = [];
    protected $fakerInstances = [];

    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $name = 'export';

    protected $dbPrefix = 'masquerade_';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Run masquerade for a specific platform and group(s)';

    /**
     * @var Config
     */
    protected $configHelper;

    /**
     * @var array
     */
    protected $fakerInstanceCache;

    /**
     * @var TableServiceFactory
     */
    private $tableServiceFactory;

    /**
     *
     */
    protected function configure()
    {
        $this
            ->setName($this->name)
            ->setDescription($this->description)
            ->addOption('config', null, InputOption::VALUE_OPTIONAL | InputOption::VALUE_IS_ARRAY, 'One or more extra config directories for config.yaml or platform configs')
            ->addOption('platform', null, InputOption::VALUE_OPTIONAL)
            ->addOption('driver', null, InputOption::VALUE_OPTIONAL, 'Database driver [mysql]')
            ->addOption('database', null, InputOption::VALUE_OPTIONAL)
            ->addOption('username', null, InputOption::VALUE_OPTIONAL)
            ->addOption('password', null, InputOption::VALUE_OPTIONAL)
            ->addOption('port', null, InputOption::VALUE_OPTIONAL, 'Database port [3306]')
            ->addOption('host', null, InputOption::VALUE_OPTIONAL, 'Database host [localhost]')
            ->addOption('prefix', null, InputOption::VALUE_OPTIONAL, 'Database prefix [empty]')
            ->addOption('locale', null, InputOption::VALUE_OPTIONAL, 'Locale for Faker data [en_US]')
            ->addOption('group', null, InputOption::VALUE_OPTIONAL, 'Which groups to run masquerade on [all]')
            ->addOption('charset', null, InputOption::VALUE_OPTIONAL, 'Database charset [utf8]')
            ->addOption('with-integrity', null, InputOption::VALUE_NONE, 'Run with foreign key checks enabled')
            ->addOption('batch-size', null, InputOption::VALUE_REQUIRED, 'Batch size to use for anonymization', 500)
            ->addOption('save-to', null, InputOption::VALUE_OPTIONAL, 'Add export location path', '.');
    }

    /**
     * Execute the console command.
     *
     * @return int
     */
    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $this->input = $input;
        $this->output = SymfonyOutput::createFromSymfonyOutput($output);

        $this->setup();

        $output->writeln(self::LOGO);
        $output->writeln('                        v' . self::VERSION);

        $startTime = new \DateTime();

        foreach ($this->config as $groupName => $tables) {
            if (!empty($this->group) && !in_array($groupName, $this->group)) {
                continue;
            }
            foreach ($tables as $tableName => $table) {
                $table['name'] = $tableName;
                $this->fakeData($table);
            }
        }

        $this->output->success('Anonymization complete in [%s]', $startTime->diff(new \DateTime())->format('%h:%i:%s'));
        $this->output->success('Db exported in '.rtrim($this->input->getOption('save-to'),'/').'/bk.sql');
        $this->export();
        return 0;
    }

    private function export()
    {
        $databaseConfig = $this->configHelper->readConfigFile();
        $tempDbName = $this->$dbPrefix . $databaseConfig['database'];
        exec('mysqldump -u '.$databaseConfig['username'].' -p'.$databaseConfig['password'].' '.$tempDbName.' > '.rtrim($this->input->getOption('save-to'),'/').'/bk.sql');
        $this->db->statement('DROP DATABASE IF EXISTS '.$tempDbName.';');
        //todo remove user permisions over temporary databse that is beeing deleted.
    }

    /**
     * @param array $table
     */
    private function fakeData(array $table) : void
    {
        $table['provider'] = $table['provider'] ?? [];

        if (is_string($table['provider'])) {
            $this->output->errorAndExit(
                'Provided configuration "%s" is not compatible with new version of masquerade, please use processor_factory instead.',
                $table['provider']
            );
        }

        $dataProcessorFactoryClass = $table['processor_factory'] ?? self::DEFAULT_DATA_PROCESSOR_FACTORY;

        if (!class_exists($dataProcessorFactoryClass)) {
            $this->output->errorAndExit(
                'Provided %s class does not exists.',
                $dataProcessorFactoryClass
            );
        }

        if (!in_array(DataProcessorFactory::class, class_implements($dataProcessorFactoryClass))) {
            $this->output->errorAndExit(
                'Provided %s class does not implement required %s interface.',
                $dataProcessorFactoryClass,
                DataProcessorFactory::class
            );
        }

        /** @var DataProcessorFactory $dataProcessorFactory */
        $dataProcessorFactory = new $dataProcessorFactoryClass();
        $tableName = $table['name'];
        try {
            $dataProcessor = $dataProcessorFactory->create($this->output, $this->tableServiceFactory, $table);
        } catch (TableDoesNotExistsException $exception) {
            $this->output->info('Table %s does not exists. Skipping...', $tableName);
            return;
        }


        $this->output->debug(
            'Updating table using the following data',
            [
                'processor' => get_class($dataProcessor),
                'configuration' => $table
            ]
        );

        $isIntegrityImportant = $this->input->hasOption('with-integrity') || $table['provider']['where'] ?? '';
        $isDelete = $table['provider']['delete'] ?? false;
        $isTruncate = $table['provider']['truncate'] ?? false;

        if ($isIntegrityImportant && $isDelete) {
            $this->output->info('Deleting records from %s table', $tableName);
            $dataProcessor->delete();
            $this->output->success('Records have been deleted from %s table', $tableName);
            return;
        } elseif ($isDelete || $isTruncate) {
            $this->output->info('Truncating records from %s table', $tableName);
            $dataProcessor->truncate();
            $this->output->success('Records have been truncated from %s table', $tableName);
            return;
        }

        try {
            $dataProcessor->updateTable((int)$this->input->getOption('batch-size'), \Closure::fromCallable([$this, 'generateRecord']));
        } catch (\Exception $e) {
            $this->output->errorAndExit($e->getMessage());
        }

        $this->output->info('');
    }

    private function generateRecord(iterable $columns): \Generator
    {
        foreach ($columns as $columnData) {
            $formatter = $columnData['formatter']['name'] ?? null;
            $formatterData = $columnData['formatter'] ?? [];
            $providerClassName = $columnData['provider'] ?? false;

            if (!$formatter) {
                $formatter = $formatterData;
                $options = [];
            } else {
                $options = array_values(array_slice($formatterData, 1));
            }

            if (!$formatter) {
                yield null;
                continue;
            }

            if ($formatter == 'fixed') {
                yield Arr::first($options);
                continue;
            }

            try {
                $fakerInstance = $this->getFakerInstance($columnData, $providerClassName);
                if ($columnData['unique'] ?? false) {
                    $fakerInstance = $fakerInstance->unique();
                } elseif ($columnData['optional'] ?? false) {
                    $fakerInstance = $fakerInstance->optional();
                }

                yield $this->asScalar($fakerInstance->{$formatter}(...$options));
            } catch (\InvalidArgumentException $e) {
                // If InvalidArgumentException is thrown, formatter is not found, use null instead
                yield null;
            }
        }
    }

    /**
     * Transforms non scalar values to scalar ones
     *
     * @return scalar
     */
    private function asScalar($value)
    {
        if (is_scalar($value) || $value === null) {
            return $value;
        }

        if ($value instanceof \DateTime) {
            return $value->setTimezone(new \DateTimeZone('UTC'))->format('Y-m-d H:i:s');
        }

        $this->output->debug('Unexpected type', ['value' => $value]);
        $this->output->errorAndExit('Unknown type has been provided from generator');
    }

    /**
     * @throws \Exception
     */
    private function setup()
    {
        $this->configHelper = new Config($this->input->getOptions());

        $databaseConfig = $this->configHelper->readConfigFile();

        $this->platformName = $this->input->getOption('platform') ?? $databaseConfig['platform'] ?? null;

        if (!$this->platformName) {
            $this->output->errorAndExit('No platformName set, use option --platform or set it in ' . Config::CONFIG_YAML);
        }

        $this->config = $this->configHelper->getConfig($this->platformName);

        $host = $this->input->getOption('host') ?? $databaseConfig['host'] ?? 'localhost';
        $driver = $this->input->getOption('driver') ?? $databaseConfig['driver'] ?? 'mysql';
        $database = $this->input->getOption('database') ?? $databaseConfig['database'] ?? null;
        $username = $this->input->getOption('username') ?? $databaseConfig['username'] ?? null;
        $password = $this->input->getOption('password') ?? $databaseConfig['password'] ?? null;
        $prefix = $this->input->getOption('prefix') ?? $databaseConfig['prefix'] ?? '';
        $charset = $this->input->getOption('charset') ?? $databaseConfig['charset'] ?? 'utf8';
        $port = $this->input->getOption('port') ?? $databaseConfig['port'] ?? 3306;

        $errors = [];
        if (!$host) {
            $errors[] = 'No host defined';
        }
        if (!$database) {
            $errors[] = 'No database defined';
        }
        if (!$username) {
            $errors[] = 'No username defined';
        }
        if (count($errors) > 0) {
            $this->output->errorAndExit(implode(PHP_EOL, $errors));
        }

        $capsule = new Capsule;
        $capsule->addConnection([
            'driver'    => $driver,
            'host'      => $host,
            'database'  => $database,
            'username'  => $username,
            'password'  => $password,
            'prefix'    => $prefix,
            'charset'   => $charset,
            'port'      => $port
        ]);

        $this->db = $capsule->getConnection();
        $tempDbName = $this->$dbPrefix . $database;
        $this->db->statement("GRANT ALL PRIVILEGES ON ".$tempDbName.".* TO '".$username."'@'localhost';");
        $this->db->statement('DROP DATABASE IF EXISTS '.$tempDbName.';');
        $this->db->statement('CREATE DATABASE '.$tempDbName.';');
        exec('mysqldump -u '.$username.' -p'.$password.' '.$database.' | sed "s/\`'.$database.'\`./\`'.$tempDbName.'\`./" | mysql -u '.$username.' -p'.$password.' -D '.$tempDbName);

        $capsule = new Capsule;
        $capsule->addConnection([
            'driver'    => $driver,
            'host'      => $host,
            'database'  => $tempDbName,
            'username'  => $username,
            'password'  => $password,
            'prefix'    => $prefix,
            'charset'   => $charset,
            'port'      => $port
        ]);

        $this->db = $capsule->getConnection();
        if (!$this->input->getOption('with-integrity')) {
            $this->output->info('[Foreign key constraint checking is off - deletions will not affect linked tables]');
            $this->db->statement('SET FOREIGN_KEY_CHECKS=0');
        }

        $this->db->statement("SET SESSION sql_mode=''");

        $this->tableServiceFactory = new TableServiceFactory($this->db);

        $this->locale = $this->input->getOption('locale') ?? $databaseConfig['locale'] ?? 'en_US';

        $this->group = array_filter(array_map('trim', explode(',', $this->input->getOption('group'))));
    }

    /**
     * @param array $columnData
     * @param bool $providerClassName
     * @return Generator
     * @throws \Exception
     * @internal param bool $provider
     */
    private function getFakerInstance(array $columnData, $providerClassName = false) : Generator
    {
        $key = md5(serialize($columnData) . $providerClassName);
        if (isset($this->fakerInstanceCache[$key])) {
            return $this->fakerInstanceCache[$key];
        }

        $fakerInstance = FakerFactory::create($this->locale);

        $provider = false;
        if ($providerClassName) {
            $provider = new $providerClassName($fakerInstance);
        }

        if (is_object($provider)) {
            if (!$provider instanceof \Faker\Provider\Base) {
                $this->output->errorAndExit('Class ' . get_class($provider) . ' is not an instance of \Faker\Provider\Base');
            }
            $fakerInstance->addProvider($provider);
        }

        $this->fakerInstanceCache[$key] = $fakerInstance;

        return $fakerInstance;
    }
}
